### PR TITLE
Solaraviz api

### PR DESCRIPTION
**Pull request by Corvince**
This is the start of a new API for the Solara frontend. See https://github.com/projectmesa/mesa/discussions/2236 for more background information. I will eventually update this intro to accommodate more information, but I wanted to start this draft PR now and not wait longer just because I am too lazy right now to write more sentences.

So far everything worked out nicely. The interesting changes are in the `myviz.py`  file. This is where the new entrypoint can be found. The other components ModelController and UserParams only received very minor changes and were just moved to other files to not interefere with the current SolaraViz (during development).

MyViz also includes a minimal example, if someone wants to try this out just run `solara run myviz.py`

image

### CODE REVIEW

**Comment by EwoutH** [comment]
Exciting!

I tried both
```bash
mesa>solara run mesa/visualization/myviz.py
```
and 
```bash
mesa\mesa\visualization>solara run myviz.py
```
But for some reason I can't get the imports working. Will take a better look tomorrow.

**Comment by rht** [comment]
I tried it and it worked. Will add inline comments.

File: mesa/visualization/myviz.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            self.grid.place_agent(
+                a, (self.random.randrange(10), self.random.randrange(10))
+            )
+
+    def step(self):
+        self.agents.do("step")
+
+
+model = MyModel(10)
+Page = MyViz(model, [ShowSteps, ShowNumAgents, make_space_altair()])
```
Comment (rht) [inline mesa/visualization/myviz.py:97]: New API looks more modular. This means user can add any existing Solara components from Solara library, including Plotly if they want to.

I think it would be convenient to have a string "default" that automatically expands into `[ShowSteps, make_space_altair()]`. Not sure about num agents being shown by default, but this is a minor point.

**Comment by EwoutH** [comment]
@rht from where did you run it? I would also like to try it.

File: mesa/visualization/utils.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import solara
+
+_update_counter = solara.reactive(0)
```
Comment (rht) [inline mesa/visualization/utils.py:3]: Is this global variable a temporary artifact? Because we want to support showing multiple SolaraViz in one go: https://simrisk.pulcloud.io/.

**Comment by rht** [comment]
I ran it In the visualization directory.

**Comment by EwoutH** [comment]
Maybe something with working directories? Hate this kind of stuff, and imports in general. So many inexplainable issues.

image

**Comment by EwoutH** [comment]
It was a namespace conflict between the installed Mesa package and the GitHub repo. Removing mesa and adding

```Python
import sys
import os

sys.path.insert(0, os.path.abspath('C:/Users/Ewout/Documents/GitHub/mesa'))
```
solved it. Let's take a look!

**Comment by EwoutH** [comment]
Looks clean, feels fast and responsive!

What's the difference between these two buttons? When running they both seem to pauze the simulation in the same way.

image

File: mesa/visualization/ModelCreator.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
+import solara
+
+from mesa.visualization.solara_viz import split_model_params
+from mesa.visualization.UserParam import Slider
+
+
+@solara.component
+def ModelCreator(model, model_cls, model_params):
```
Comment (EwoutH) [inline mesa/visualization/ModelCreator.py:8]: So you can input a model instance, model class and params? I get either the first or the second two are required?

If you input an instance, will it be updated in place, so you can access it from your interpreter later?

File: mesa/visualization/ModelCreator.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                min=options.min,
+                max=options.max,
+                step=options.step,
+            )
+            continue
+
+        # label for the input is "label" from options or name
+        label = options.get("label", name)
+        input_type = options.get("type")
+        if input_type == "SliderInt":
```
Comment (EwoutH) [inline mesa/visualization/ModelCreator.py:60]: What does this structure add to the user? Isn't it easier to just let them just the Solara components directly?

**Comment by rht** [comment]
The 2 buttons having the same functionality is out of scope for the PR. It is due to ipywidget displaying both and there is no way to turn either off.

**Review by EwoutH — COMMENTED** [review]
I really love that you can just input an model instance, and if I'm correct, acces that model instance from the interpreter.

This would solve a huge part of https://github.com/projectmesa/mesa/issues/2176, right?

File: mesa/visualization/ModelCreator.py (hunk 4 above)
Comment (rht) [reply mesa/visualization/ModelCreator.py:60]: This is already present in existing SolaraViz, and so is out of scope. The point was so that user can declaratively specify the param in a JSON, and so is serializable.

Comment (rht) [reply mesa/visualization/ModelCreator.py:60]: If the scope of the PR is small, the API change can be merged sooner.

Comment (EwoutH) [reply mesa/visualization/ModelCreator.py:60]: Thanks for the context, I’m not 100% up to speed with the current visualization.

**Comment by Corvince** [comment]
I have updated the code so it now updates solara_viz.py instead of creating a new myviz. Also moved back the modified ModelController.

All in all the diff should be much more reflective of the actual code changes. 

Things that are still missing (please add if you notice something):

- [ ] Layout stuff (draggable grid layout)
- [ ] reactive seed
- [ ] Easy way to plot simple measures just by their attribute name

File: mesa/visualization/myviz.py (hunk 1 above)
Comment (Corvince) [reply mesa/visualization/myviz.py]: Yes I think thats a good idea. Num agents are currently just shown for testing purpose - I don't think it should be default.

File: mesa/visualization/utils.py (hunk 2 above)
Comment (Corvince) [reply mesa/visualization/utils.py]: I know that has been an issue in older versionis of solara, but solara might handle this gracefully. The docs now include this sentence

> Reactive objects in Solara are also context-aware, meaning that they can maintain separate values for each browser tab or user session. This enables each user to have their own independent state, allowing them to interact with the web application without affecting the state of other users.
https://solara.dev/documentation/api/utilities/reactive

But I haven't tested his yet.

**Comment by Corvince** [comment]
> I really love that you can just input an model instance, and if I'm correct, acces that model instance from the interpreter.
> 
> This would solve a huge part of https://github.com/projectmesa/mesa/issues/2176, right? And would [this](https://github.com/projectmesa/mesa/issues/2176#issuecomment-2282808460) be possible?

Yes, exactly! Although for it to fully work, even if you reset the model or change parameters, you should pass in the model as a reactive model, i.e. call `solara.reactive(model)` on it. Inside the SolaraViz it also attaches a call to `force_update()` after each step, which triggers a viz update.

File: mesa/visualization/ModelCreator.py (hunk 3 above)
Comment (Corvince) [reply mesa/visualization/ModelCreator.py]: Actually it expects a reactive version of a model, so that the model from the outside still receives the recreated versions. 

I now removed the model_cls, as it can be resolved from the model instance. 

So now it receives an actual model (as a reactive variable) and user params. So I think ModelCreator isn't a good name

**Comment by EwoutH** [comment]
If you need any input let me know!

File: mesa/visualization/utils.py (hunk 2 above)
Comment (rht) [reply mesa/visualization/utils.py]: What about the case when there are multiple `page` objects in a single Jupyter notebook? As can be seen in the [visualization tutorial](https://mesa.readthedocs.io/en/latest/tutorials/visualization_tutorial.html). They seem to share the same context.

Comment (Corvince) [reply mesa/visualization/utils.py]: That shouldn't be a big problem luckily. It only forces a rerender, no data change. So at worst you have a few unnecessary rerenders in your notebook. Let's wait and see if this causes serious performance issues and if yes reconsider. But for now it should be good

**Comment by Corvince** [comment]
Now that you say it, what's your opinion on the layout? I thought at first to implement how it's currently done. But maybe we can be more unopinionated. In my implementation all components are just displayed with no surrounding layout. But the components themselves can of course include any layout. We could keep it that way and provide some layout components. So users could just pass `solaraviz(model, components=GridDraggable(componentA, componentB)` or something like that.

Also, regarding naming. Is Components a good enough name?

**Comment by EwoutH** [comment]
I think the current layout is quite good. I like controls on the left, visualisation and output in the center/right.
- Maybe it would be nice if you could specify is you want things below or right of previous components, but maybe keeping it adaptive is more flexible.
- It might be nice to be able to specify an aspect ratio per component. Some graphs you might want wide, some tall etc.
  - For the `space_drawer` this might be done automatically based on the grid width and height.

> Also, regarding naming. Is Components a good enough name?

Many names could work here, but Components is quite good.

**Comment by rht** [comment]
I actually find the naming to be quite restrictive in order to encompass other Solara components, because it needs to be general, e.g. if you want to use the `DataFrame`/`PivotTable` component in https://solara.dev/documentation/components. "Measures" doesn't cover this. Meanwhile, "components" is what is already used by Solara.

File: mesa/visualization/utils.py (hunk 2 above)
Comment (rht) [reply mesa/visualization/utils.py]: I also notice that all the components take `model` as an argument, and so, would it be possible to smuggle in the update counter as an attribute of the model?

Comment (Corvince) [reply mesa/visualization/utils.py]: Interesting idea. Not sure this will work. I am rather sceptical, but will definitely try it out.

**Comment by quaquel** [comment]
This is just a clarifying question from my end. In the various conversations on data collection, we converged on building this on top of some pub-sub or signal-based API. How might that choice affect this effort, or vice versa, what would be convenient to use inside MESA such that both data collection and model visualization can utilize it?

In #2176, @maartenbreddels suggested using either Solara reactive variables or psygnal. I have looked at the latter, and it is very elegant and does a lot of what we want for data collection. Importantly, it also allows for subscribing to updates in collections. However, it would add an additional dependency to MESA. I also started exploring Solara's reactive variables. It appears this covers a subset of what we would need for data collection, but I am not sure until I actually play around with it more. 

In my vision, both the visualization side and the data collection side of MESA use the same mechanism for being informed/getting state information from a MESA model.

**Comment by Corvince** [comment]
> This is just a clarifying question from my end. In the various conversations on data collection, we converged on building this on top of some pub-sub or signal-based API. How might that choice affect this effort, or vice versa, what would be convenient to use inside MESA such that both data collection and model visualization can utilize it?
> 
> In #2176, @maartenbreddels suggested using either Solara reactive variables or psygnal. I have looked at the latter, and it is very elegant and does a lot of what we want for data collection. Importantly, it also allows for subscribing to updates in collections. However, it would add an additional dependency to MESA. I also started exploring Solara's reactive variables. It appears this covers a subset of what we would need for data collection, but I am not sure until I actually play around with it more.
> 
> In my vision, both the visualization side and the data collection side of MESA use the same mechanism for being informed/getting state information from a MESA model.

The top level view on how it currently works (in this PR) is that each component registers a "update_counter" as a render dependency. Whenever this counter increases the component rerenders. Currently the counter is increased after each mode..step.

Of course it would be much more elegant to auto-update whenever an attribute changes, for example not through a call to step, but maybe through some form of user interactions. If the variable would be a reactive variable this would already happen automatically. 

So from a visualization point of view the use of solaras reactive values would be very charming. So I would appreciate it if you further investigate solaras reactive suitability. 

But overall this PR is mostly neutral to how we handle this - it mostly deals with the surroundings and establishing a cleaner API. The use of signals/reactive values/pub-sub will mostly play some role for individual components (e.g. a plot of a specifc variable)

**Comment by Corvince** [comment]
I just marked this as ready for review! Even though the tests are still failing, this should probably also be fixed as part of this PR? 

Otherwise I think this is good to review as a first PR just for the API changes.

Next todos are:

- [ ] Update example models
- [ ] Write migration guide from current API to this PR. You should be able to have the same viz as before with small adjustemnts, but those adjustments need to be written out.
- [ ] Bring back GridDraggable layout. This was used before, but I found the code a bit confusing so I didn't manage to adapt just yet. Per default the layout now is unopinionated, but you can bring back the old style with `SolaraViz(model, components=[GridDraggable(...components)]`. But the GridDraggable component needs to be written in a further PR
- [ ] Decide which subcomponents will be part of the public API for Mesa v3. 

Not so immediate, but also a nice todo: Rewrite ModelController. There is already code for a much simpler threaded version and from a first test this finally seems to work. Maybe it doesn't even need to be threaded and just works. In any case it can be vastly simplified (and no more unnecessary pause and stop button)

But first lets get this PR through!

**Comment by EwoutH** [comment]
@Corvince looks awesome! Any specific things you would like me to test or review?

**Comment by Corvince** [comment]
Yes! You can try to use it 🙈 

Heres a quick migration guide. Where previous you could have something like this for a minimal frontend
```python
SolaraViz(model_cls, model_params, agent_portrayal=agent_portrayal)
```

you would now do

```python
SolaraViz(model, components=[make_space_altair(agent_portrayal)])
```

you can test if that works for `make_space_altair` and `make_space_matplotlib` and then additionally add `model_params` to SolaraViz. And then test if everything works (start, stop, reset, reseed, change parameters).

That would be awesome!

**Comment by EwoutH** [comment]
Awesome, I'm going to 
```
pip install -U -e git+https://github.com/projectmesa/mesa@solaraviz-api#egg=mesa
```
and test some examples.

I just talked to Jan, he is going to try to dust of hist pub-sub efforts to see if that will integrate nicely with this.

**Comment by EwoutH** [comment]
Playing a bit around! Got `boltzmann_wealth_model_experimental` working:

image

> * Easy way to plot simple measures just by their attribute name

Yeah this would be really useful. We could think of another/more robust API to parse measures other than just a string name (@quaquel?), but I think this is important to get right.

I like that's way more flexible: You don't need to input `model_params` for values you don't need as sliders and such. You can just use the default values or override them.

Another this that's interesting is that the slider overrides default/other values. I think that's good to document.

image

It does show that something like 
- https://github.com/projectmesa/mesa/issues/2268

can be useful.

**Comment by EwoutH** [comment]
I also tried [`virus_on_network`](https://github.com/projectmesa/mesa-examples/tree/main/examples/virus_on_network). However, it has a custom [`make_plot`](https://github.com/projectmesa/mesa-examples/blob/330b4d9c8e44127b3c4b89603972a46dd5ec39b9/examples/virus_on_network/app.py#L47C1-L60C33) function, that seems difficult to adapt, and I didn't get it to work.

```Python
def make_plot(model):
    # This is for the case when we want to plot multiple measures in 1 figure.
    fig = Figure()
    ax = fig.subplots()
    measures = ["Infected", "Susceptible", "Resistant"]
    colors = ["tab:red", "tab:green", "tab:gray"]
    for i, m in enumerate(measures):
        color = colors[i]
        df = model.datacollector.get_model_vars_dataframe()
        ax.plot(df.loc[:, m], label=m, color=color)
    fig.legend()
    # Set integer x axis
    ax.xaxis.set_major_locator(MaxNLocator(integer=True))
    solara.FigureMatplotlib(fig)
```

```Python
page = SolaraViz(
    VirusOnNetwork,
    model_params,
    measures=[
        make_plot,
        make_text(get_resistant_susceptible_ratio),
    ],
    name="Virus Model",
    agent_portrayal=agent_portrayal,
)
page  # noqa
```

**Comment by Corvince** [comment]
Awesome, thank you so much for testing this out!

> > * Easy way to plot simple measures just by their attribute name
> 
> Yeah this would be really useful. We could think of another/more robust API to parse measures other than just a string name (@quaquel?), but I think this is important to get right.

Yes, this needs some rework. Currently it takes its data from the data collector, but I think it would make more sense to get the values directly. But I already implemented the current approach with `make_plot_measure("foo")`


> I also tried [`virus_on_network`](https://github.com/projectmesa/mesa-examples/tree/main/examples/virus_on_network). However, it has a custom [`make_plot`](https://github.com/projectmesa/mesa-examples/blob/330b4d9c8e44127b3c4b89603972a46dd5ec39b9/examples/virus_on_network/app.py#L47C1-L60C33) function, that seems difficult to adapt, and I didn't get it to work.

It should be as easy as turning it into a solara component and adding the update_counter:

```python
@solara.component  # first change
def make_plot(model):
    update_counter.get() # second change
    # This is for the case when we want to plot multiple measures in 1 figure.
    fig = Figure()
    ax = fig.subplots()
    measures = ["Infected", "Susceptible", "Resistant"]
    colors = ["tab:red", "tab:green", "tab:gray"]
    for i, m in enumerate(measures):
        color = colors[i]
        df = model.datacollector.get_model_vars_dataframe()
        ax.plot(df.loc[:, m], label=m, color=color)
    fig.legend()
    # Set integer x axis
    ax.xaxis.set_major_locator(MaxNLocator(integer=True))
    solara.FigureMatplotlib(fig)
```

But yes, documenting how to add author custom components should als be pretty high on the follow up todo list

**Comment by EwoutH** [comment]
The only question I now have, is do we want to directly replace the current viz or do we want to have some deprecation scheme and have them parallel for a bit?

**Comment by maartenbreddels** [comment]
> this could be done using Solara's `Reactive`, but this is closely wrapped up with all kinds of GUI stuff that we don't need.

Why do you think that is the case? They are, I think, quite separated from the UI part (except being in the same package).

Something to also consider is 'scope'. In solara server (if you plan to use that), you can have multiple browser pages (often seen as 'users') connected to the same runtime. Reactive variables understand that, similar to threading.local for thread local storage, Solara reactive variables use kernel-local storage.

**Comment by quaquel** [comment]
> Why do you think that is the case? They are, I think, quite separated from the UI part (except being in the same package).

Let me clarify. I have been trying to figure out which parts from Solara I can use to reimplement my existing pubsub idea and envisioned API. The documentation for Solara focusses on "GUI stuff". So, I have been diving into the source code. This has taken me to parts of the code base that are not documented (e.g.,`Reactive`), so I have been trying to figure out what is going on and how to utilize it for MESA purposes.

 In short, my impression is based perhaps more on the existing Solara documentation and my struggle to parse from the entire code base the parts I would need. As you said yourself in #2176, ideally the reactive part should be its own library.

**Comment by Corvince** [comment]
> The only question I now have, is do we want to directly replace the current viz or do we want to have some deprecation scheme and have them parallel for a bit?

Thats a good question and one I have been pondering for a while. Technically the solaraviz has always been experimental and so we don't break any contract with the changes. However, there hasn't been an alternative built-in frontend for a while now and we have been pointing towards the new viz for some time. It will be a bit annoying for everyone to rebuild there modesl yet again. But then again we have several breaking changes with mesa 3.0, so people will probably have to rebuild their models anyway. 

One difficulty with deprecation is how to handle it gracefully. I think the code for that will be a bit ugly and then is it really necessary to deprecate experimental features? The most elegant deprecation would of course be to find a new name for "SolaraViz". I think its not perfect at all, because its neither specific to mesa, nor generic enough. Sure, currently it only supports solara components, but maybe this could change in the future (e.g. you can already use vue components, but you need to wrap them with `@solara.vue_component`). 

So maybe there is a better name to be discovered? Then we could leave the old code intact (deprecated) and write nice migration guides to the new viz.

**Comment by EwoutH** [comment]
Since we haven't clear policy, versioning systems and integration testing in our examples, I would propose placing this in the `.experimental` space for now. That way we can continue to develop this at a faster pace without being afraid that we have to update everything continuously.

Furthermore, I really want to be able to test the visualisation in CI. I think this will really help with that:

> I want to directly trigger run the Solara visualisation from the Python script. So doing something like `page.run()` or whatever.

Would something like this be possible?

**Comment by EwoutH** [comment]
> One difficulty with deprecation is how to handle it gracefully.

1. Add new stuff in new location
2. Develop it further
3. Declare it stable
4. Start pointing to it, write docs, etc.
5. Start deprecating the old one (throw a warning)
6. Wait a few versions
7. Throw error for the old one
8. Wait a few versions
9. Remove the old one
10. If preferred, move the new one. Point from the initial new location to the new initial location not to break things.

Best for it would be to have a new name to make it clearly distinct.

**Comment by Corvince** [comment]
> > I want to directly trigger run the Solara visualisation from the Python script. So doing something like `page.run()` or whatever.

I am not sure I understand this correctly. Whats the workflow here? Normally you would execute `solara run file.py`. Then you can edit the file and solara will reload and reflect all changes. 
Or you can use the solaraviz in a notebook. Then (with this PR) if you pass in a reactive model you can modify the model outside the solaraviz display, and it will update anytime you call `model.value.force_update()`. This is automatically done for `model.step`

Or do you mean in an interactive python session? Not sure if this is possible, but since solara automatically reloads, I think its easier to just start it on a file.

**Comment by EwoutH** [comment]
I just want to press run here (which does `python app.py`) and have the visualisation start.

image

**Comment by Corvince** [comment]
I think you just need some custom settings in your editor. Have a look at

https://solara.dev/documentation/advanced/howto/debugging

**Comment by EwoutH** [comment]
I tried a lot of things but unfortunately none worked.

The point it, I just want to run a Python script when I'm normally do, and then, when the SolaraViz instance is called (`page()`, `page.run()`, whatever), Solara starts. Then, I do my stuff in Solara, at some point I'm done, press something like "Close", and now I'm back in the Python script and if there's any code left, it continues running.

(we can have this discussion in a separate thread if you want, since it's not directly linked to this PR)

**Comment by Corvince** [comment]
> Since we haven't clear policy, versioning systems and integration testing in our examples, I would propose placing this in the `.experimental` space for now. That way we can continue to develop this at a faster pace without being afraid that we have to update everything continuously.

Disagree here. We currently don't have a frontend currently. The latest stable release (2.3) still has the "old" SolaraViz under experimantal. The next release (3.0) will have the "old" solaraviz marked as stable. But the whole purpose of this PR was to have a good, stable API for 3.0. Dont like the idea of moving this to experimental and then soon deprecating the "old" SolaraViz, when its not out of experimental yet.

**Comment by EwoutH** [comment]
Yeah that's a good point. I'm inclined to agree.

But then the docs need to be on-par before we get into Mesa 3.0 beta territory. Do we have the capacity to do that (@projectmesa/maintainers)?

**Comment by rht** [comment]
> But then the docs, tutorial and examples need to be on-par before we get into Mesa 3.0 beta territory. Do we have the capacity to do that (@projectmesa/maintainers)?

It may seem to be a lot of work, but given that in Mesa 2.x it is still an experimental API, users of the JupyterViz (now SolaraViz) were already expecting API changes. It's less work overall to include the API changes altogether now, than to delay it to Mesa 4.0.

**Review by EwoutH — APPROVED** [review]
I'm good with it. Just saying it has to be done.

**Comment by Corvince** [comment]
Exactly. I think it may make sense to actually move the JupyterViz/SolaraViz back into experimental. This way 2.x users who already use the experimental SolaraViz module don't *need* to change anything. They can still `import SolaraViz from mesa.experimental`  But at the same time we have the new, stable API `import SolaraViz from mesa.visualization`.  How does that sound? This way we only need changes from people on the 3.x pre release channel, but changes are expected there. 

Still we should consider a new name, but this can come later (before 3.0 stable of course. But if we have a new name we can alias it to SolaraViz, like we currently do for JupyterViz)

**Comment by Corvince** [comment]
I rebased this PR. Hopefully tonight or tomorrow at the latest I will update the tests, so ci will be happy again

**Comment by EwoutH** [comment]
Leftover from #2278, but could you fix this import?

https://github.com/projectmesa/mesa/blob/cfa26359b3aecef12636031e08230d38768ae4b9/tests/test_solara_viz.py#L8

(I already approved)

**Comment by Corvince** [comment]
Yes, see my last comment

**Comment by wang-boyu** [comment]
> A new name for the new viz would help a lot. Maybe just `MesaViz`?

How about `ModelViz` or `ModelView`?

**Comment by Corvince** [comment]
I would like to extend the functionality to allow for more model interactions, adding plots interactively and having an "agent observer" for single agents. So not just a pure visualization. After a chat with chatgpt my best idea so far might be "ModelExplorer"

**Comment by EwoutH** [comment]
`ModelExplorer` sounds good. Also fresh, and new.

I would everything that's strictly in Solara (like Solara-components) to keep a Solara-like name. Then if we add non-Solara stuff, we can give them non-Solara names.

**Comment by wang-boyu** [comment]
Hi @tpike3, we probably need to rethink the jupyter visualization for mesa-geo, considering the new API design from this PR.

**Comment by Corvince** [comment]
Can you point me to that? I can take a look tomorrow

**Comment by wang-boyu** [comment]
Thanks! The current implementation is here: https://github.com/projectmesa/mesa-geo/tree/main/mesa_geo/visualization, from PR https://github.com/projectmesa/mesa-geo/pull/212.

There was some comments from @rht that needs to be addressed: https://github.com/projectmesa/mesa-geo/pull/199#issuecomment-2041764285 but we decided to merge and have something working first.

As I mentioned above in https://github.com/projectmesa/mesa/pull/2263#discussion_r1743872440, we probably don't need to copy over anything. Instead, we could try to refactor existing implementations into some solara components for geospace.

Let me know if you need anything else : ) also cc @tpike3

**Comment by Corvince** [comment]
Thanks for the extensive reply! Sorry, I didn't catch your previous comment, but you are spot on! 

It will just require a function `make_geospace_leaflet()` that returns what I think is the "map" component currently. So something like this

```python
from mesa.visualization.utils import update_counter

def make_geospace_leaflet(agent_portrayal, view, zoom, tiles,scroll_wheel_zoom):

    map_drawer = leaflet_viz.MapModule(
        portrayal_method=agent_portrayal,
        view=view,
        zoom=zoom,
        tiles=tiles,
        scroll_wheel_zoom=scroll_wheel_zoom,
    )

    center_default= ...


    @solara.component
    def map(model):
        update_counter.get() # <- Important line to notify the frontend whenever model updates (e..g step)
        # render map in browser
        zoom_map = solara.reactive(zoom)
        center = solara.reactive(center_default)

        base_map = map_drawer.tiles
        layers = map_drawer.render(model)

        ipyleaflet.Map.element(
            zoom=zoom_map.value,
            center=center.value,
            scroll_wheel_zoom=scroll_wheel_zoom,
            layers=[
                ipyleaflet.TileLayer.element(url=base_map["url"]),
                ipyleaflet.GeoJSON.element(data=layers["agents"][0]),
                *layers["agents"][1],
            ],
        )

    return map
```

Then in a notebook you should already be able to do 
```python
GeoSpaceLeaflet = make_geospace_leaflet(...)

GeoSpaceLeaflet(model)
```

And you should see your visualization! And then you can of course pass it to SolaraViz 

Would you like to give it a shot? Unfortunately I don't have the mesa_geo stack currently installed and last time it was a bit of a hassle to get the geo tools to work under windows. Just ask if you need any help

**Comment by wang-boyu** [comment]
Thanks! This is precisely what I thought. I will find some time to try it out!

**Comment by EwoutH** [comment]
I would prefer to have a new name before a potential (alpha) release.

**Comment by EwoutH** [comment]
> I would prefer to have a new name before a potential (alpha) release.

> `ModelExplorer` sounds good. Also fresh, and new.

What do we think of this?

<hr>

On another note, te [Visualization tutorial](https://mesa.readthedocs.io/en/latest/tutorials/visualization_tutorial.html), since it’s still based on the previous version of SolaraViz. I see two options:
1. Use the previous SolaraViz version for now, which we moved to experimental.
2. Update to the new API.

**Comment by Corvince** [comment]
> On another note, te [Visualization tutorial](https://mesa.readthedocs.io/en/latest/tutorials/visualization_tutorial.html), since it’s still based on the previous version of SolaraViz. I see two options:
> 
> 1. Use the previous SolaraViz version for now, which we moved to experimental.
> 2. Update to the new API.

I can quickly do a minimal update to the new API and then do a bit of an overhaul later on, if thats fine?

**Comment by EwoutH** [comment]
Sounds perfect!

**Comment by jackiekazil** [comment]
I am late, but I like ModelExplorer
